protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunction.java

```diff
@@ -0,0 +1,172 @@
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.annotation.ResponseConverterFunction;
+
+/**
+ * A {@link ResponseConverterFunction} which creates an {@link HttpResponse} with
+ * {@code content-type: application/protobuf} or {@code content-type: application/json; charset=utf-8}.
+ * If the returned object is instance of {@link MessageLite}, the object can be converted to
+ * <a href="https://developers.google.com/protocol-buffers/docs/encoding">Protocol Buffers</a> or
+ * <a href="https://developers.google.com/protocol-buffers/docs/proto3#json">JSON</a> format.
+ *
+ * <p>The Protocol Buffers spec does not have an official way to sending multiple messages because
+ * an encoded message does not have self-delimiting.
+ * See <a href="https://developers.google.com/protocol-buffers/docs/techniques#streaming">Streaming Multiple Messages</a>
+ * for more information.
+ * Therefore if the returned object is instance of {@link Publisher}, {@link Stream} or {@link Iterable}
+ * which produces {@link MessageLite}s, the object is only able to convert to
+ * <a href="https://tools.ietf.org/html/rfc7159#section-5">JSON array</a> format.
+ *
+ * <p>Note that this {@link ResponseConverterFunction} is applied to the annotated service by default,
+ * so you don't have to set explicitly unless you want to use your own {@link Printer}.
+ */
+@UnstableApi
+public final class ProtobufResponseConverterFunction implements ResponseConverterFunction {
+
+    private static final Printer defaultJsonPrinter = JsonFormat.printer();
+
+    private final Printer jsonPrinter;
+
+    /**
+     * Creates an instance with the default {@link Printer}.
+     */
+    public ProtobufResponseConverterFunction() {
+        this(defaultJsonPrinter);
+    }
+
+    /**
+     * Creates an instance with the specified {@link Printer}.
+     */
+    public ProtobufResponseConverterFunction(Printer jsonPrinter) {
+        this.jsonPrinter = requireNonNull(jsonPrinter, "jsonPrinter");
+    }
+
+    @Override
+    public HttpResponse convertResponse(ServiceRequestContext ctx, ResponseHeaders headers,
+                                        @Nullable Object result, HttpHeaders trailers) throws Exception {
+        final MediaType mediaType = headers.contentType();
+        if (result instanceof MessageLite) {
+            if (mediaType != null) {
+                if (mediaType.is(MediaType.JSON) || mediaType.subtype().endsWith("+json")) {
+                    final Charset charset = mediaType.charset(StandardCharsets.UTF_8);
+                    if (charset.contains(StandardCharsets.UTF_8)) {
```

Comment:
So this is not working if the charset is not UTF_8, right?

Reply:
Maybe we can support other encodings as well?